[1단계 - 장기] 머랭 미션 제출합니다.

cookie-meringue opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [링크](https://prolog.techcourse.co.kr/studylogs)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

Controller, View, 장기 보드와 몇몇 군데를 제외하고는 전부 객체지향적으로 설계하기 위해 노력했습니다.
빠른 시일 내에 모든 리팩터링을 마칠 계획입니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->


### 1. 상속 시 부모 필드에 접근하는 방법
초반에 상속을 사용해 `Piece` 를 설계했을 때에 드는 의문이 있었습니다.
`Piece` 는 추상 클래스로, `Position` 을 필드로 가지고 있는 구조였습니다.
```
abstract class Piece {
   protected Position position;
}
```
이 상황에서 저는 당연하게도 접근 제어자를 `protected` 로 설계했습니다. 자식 클래스에서 `Position` 에 접근할 수 있어야 했기 때문입니다.
하지만, 이런 의문이 들었습니다.

만약 자식 클래스에 부모의 `Position` 을 `null` 로 만드는 등의 잘못된 접근을 하면 위험하지 않을까? 자식은 부모 `Position` 필드의 값만 필요한데 프로퍼티로 접근을 해야 할 이유가 뭘까?

이 고민을 한 결과, 저는 부모의 필드를 `private` 으로 설정했고, 자식에서는 `getter` 를 통해 접근하도록 설계했습니다. `Piece` 의 모든 자식들을 `position` 필드의 값을 바꿀 일은 없었기 때문입니다.

저의 이러한 생각에 대해 어떻게 생각하시는 지 궁금합니다.


### 2. 기물의 움직임을 전략 인터페이스로 추상화
각 기물의 움직임을 전략으로 표현했습니다.
우선, 각 기물의 움직임을 `MovementStrategy` 로 추상화시켰습니다.
`MovementStrategy` 는 기물이 움직일 수 있는지를 반환하는 isMoveable() 를 수행하는 인터페이스입니다.
![image](https://github.com/user-attachments/assets/bc75509c-6764-4964-bb7f-20d98bd5c48f)

`MovementStrategy` 는 두 개의 전략으로 나눠집니다.
1. `FixedMovementStrategy` (고정 움직임 전략)
- 고정 움직임 전략은 고정된 좌표의 움직임을 가지는 말들의 전략입니다.
- 예를 들어 `Elephant(상)` 은 `x 축으로 2, y 축으로 3` 을 가거나 `x 축으로 3, y 축으로 2` 만큼만 갈 수 있습니다.
2. `DynamicMovementStrategy` (동적 움직임 전략)
- 동적 움직임 전략은 고정된 이동 좌표를 가지지 않고 상황에 따라 이동 좌표가 정해지는 말들의 전략입니다.
- 예를 들어 `Cannon(포)` 은 상대 말의 위치에 따라 갈 수 있는 좌표가 달라지거나 아예 갈 수 없게 됩니다.

이 두 전략을 인터페이스와 템플릿 메서드 패턴을 통해 분리함으로써, 새로운 전략이 추가되는 경우에 유연하게 대처할 수 있도록 구현했습니다.
`FixedMovementStrategy` 와 `DynamicMovementStrategy` 은 빠른 시일 내에 추상 클래스로 변경할 예정입니다.
인터페이스를 사용하면 템플릿 메서드 패턴을 사용하는 데에 있어 허점이 존재하기 때문입니다. (인터페이스의 default 메서드를 구현체가 오버라이딩할 수 있는 문제)

현재 추상 클래스로 리팩터링 완료했습니다.

이 방식에 대해 어떻게 생각하시는 지 궁금합니다.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **새 기능**
  - 한국 장기 게임의 플레이 흐름, 조각 생성, 이동 및 캡처 규칙이 구현되었습니다.
  - 사용자 입력을 통한 게임 설정 및 진행 로직이 도입되었습니다.

- **문서**
  - 게임 구조, 규칙 및 사용법에 대한 자세한 설명이 README와 사용자 매뉴얼에 추가되었습니다.

- **테스트**
  - 게임 보드, 조각 이동 및 승리 조건 등 핵심 기능의 안정성을 검증하는 다양한 테스트가 추가되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cookie-meringue committed:
docs: 기능 구현 목록 작성

cookie-meringue committed:
feat: Application 생성

Firedrago95 committed:
feat: 기물은 진영을 가진다

Firedrago95 committed:
docs: 좌표 객체 기능 목록 추가

Firedrago95 committed:
refactor: 커스텀 어노테이션 생성

Firedrago95 committed:
feat: 위치는 가로 세로 2차원 좌표를 가진다

cookie-meringue committed:
feat: 좌표를 변경할 수 있다

cookie-meringue committed:
feat: 좌표의 범위는 세로 10, 가로 9로 제한한다

- index 는 0부터 시작한다.

cookie-meringue committed:
feat: 기물은 위치를 가진다

cookie-meringue committed:
feat: 위치 비교 기능 추가

cookie-meringue committed:
refactor: Piece 를 추상 클래스로 변경

cookie-meringue committed:
feat: 모든 종류의 기물 추가

- 구현은 드라고가 진행.

Firedrago95 committed:
feat: 기물을 현재 위치로 이동할 수 없도록 구현

Firedrago95 committed:
feat: 이동할 수 없는 위치로 움직임을 제한

Firedrago95 committed:
feat: 캐논 움직임 검증 구현

cookie-meringue committed:
feat: Elephant 의 이동 가능 규칙 구현

cookie-meringue committed:
refactor: Canon 테스트 방식 변경

- 이동 가능, 불가능 테스트를 분기.

cookie-meringue committed:
remove: 사용하지 않는 import 제거

cookie-meringue committed:
refactor: 잘못 부여된 Cannon 의 역할을 Rook 으로 변경

cookie-meringue committed:
style: 코드 형식 적용

Firedrago95 committed:
feat: 이동 위치 경로상 다른 기물이 존재할 경우 이동을 제한

Firedrago95 committed:
feat: 룩 이동 경로 검증 구현

cookie-meringue committed:
refactor: 함수 분리 및 책임 분리

Firedrago95 committed:
feat: 엘리펀트 이동 경로 검증 구현

cookie-meringue committed:
refactor: 엘리펀트의 이동 가능 경로 여부 판단 함수 분리

Firedrago95 committed:
refactor: 이동 가능한 경로가 좌표 밖에 있다면 계산되지 않는다

Firedrago95 committed:
refactor: Piece 내부 함수에서 Position을 사용하도록 변경

cookie-meringue committed:
feat: 캐논 이동 가능 위치 판단 기능 구현

cookie-meringue committed:
feat: 캐논 이동 가능 경로 판단 기능 구현

Firedrago95 committed:
refactor: Cannon 내부 메서드 분리

cookie-meringue committed:
style: 코드 포맷 적용

cookie-meringue committed:
refactor: 테스트에서 사용할 Elephant 생성 함수 추가

cookie-meringue committed:
feat: 나이트 이동 구현

Firedrago95 committed:
feat: 폰 이동 구현

cookie-meringue committed:
refactor: 폰 이동 로직 함수 분리

cookie-meringue committed:
refactor: Piece 생성자 매개변수를 원시값으로 변경

cookie-meringue committed:
feat: 가드, 킹 기본 구현

- 가드와 킹의 움직임을 현재 단계에서 구현하지 않음.
- 반드시 움직일 수 없음.

cookie-meringue committed:
feat: 각 배치 별 한나라 장기말 생성기 구현

Firedrago95 committed:
feat: 각 배치 별 초나라 장기말 생성기 구현

cookie-meringue committed:
feat: 장기판 기물 초기화 기능 구현

Firedrago95 committed:
feat: 장기판 기물 이동 구현

cookie-meringue committed:
fix: static 변수가 초기화되지 않는 문제 해결

cookie-meringue committed:
refactor: 가능한 모든 경로 대신 가고자 하는 위치의 모든 경로를 탐색하도록 변경

cookie-meringue committed:
feat: view, controller 연결

cookie-meringue committed:
fix: 잘못된 기물 배치 수정

cookie-meringue committed:
refactor: 룩이 이동할 수 있는 모든 경로 탐색 후 계산하도록 변경

cookie-meringue committed:
remove: 사용하지 않는 PieceGenerator 제거

cookie-meringue committed:
fix: 잘못된 폰의 이동 범위 수정

cookie-meringue committed:
refactor: 캐논 로직 최적화

cookie-meringue committed:
refactor: 엘리펀트 이동 로직 최적화

sinb57 commented on src/main/java/janggi/domain/piece/gererator/DefaultChoPieceGenerator.java:
```diff
@@ -0,0 +1,99 @@
+
+    private List<Piece> generateKnightElephants(KnightElephantSetting knightElephantSetting) {
+
+        if (knightElephantSetting == KNIGHT_ELEPHANT_KNIGHT_ELEPHANT) {
+            return generateKnightElephantKnightElephant();
+        }
+        if (knightElephantSetting == KnightElephantSetting.KNIGHT_ELEPHANT_ELEPHANT_KNIGHT) {
+            return generateKnightElephantElephantKnight();
+        }
+        if (knightElephantSetting == KnightElephantSetting.ELEPHANT_KNIGHT_KNIGHT_ELEPHANT) {
+            return generateElephantKnightKnightElephant();
+        }
+        return generateElephantKnightElephantKnight();
+    }
```
if문의 나열이 아니라, 다형성을 활용한 풀이는 어려울까요?

cookie-meringue replied:
다음 리뷰에 반영하겠습니다!
(이제야 어느 정도 틀이 잡혀서 그렇습니다.. 죄송합니다)

cookie-meringue replied:
반영 완료했습니다.

반영하는 과정에서 장단점을 파악할 수 있었습니다.

장점: 코드 중복이 줄어듭니다. OCP 를 준수할 수 있습니다.
단점: 말에 대한 MovementStrategy가 각 Generator 별로 세세하게 달라야 한다면, 해결할 방법이 없습니다.(구조를 변경해야 합니다)


장단점을 비교해봤을 때, 장점이 단점보다 더 크게 다가왔습니다!

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,72 @@
+    public boolean isEnd() {
+        return pieceMap.values().stream()
+                .filter(value -> value.getClass() == King.class)
+                .count() != 2;
```
2는 무슨 의미일까요? 어떻게 의미를 부여할 수 있을까요?

cookie-meringue replied:
`2`는 `Side` 의 수 입니다. 
한나라와 초나라의 `King` 이 전부 남아 있다면 게임이 끝난 것이 아니므로, `Side` 의 수인 `2`를 사용했습니다.
의미를 부여하기 위해 `Side.values().length` 로 변경했습니다.

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,72 @@
+
+    public boolean isEnd() {
+        return pieceMap.values().stream()
+                .filter(value -> value.getClass() == King.class)
```
getter를 사용하지 않고, 메시지를 던지는 방법은 어떨까요?

cookie-meringue replied:
메시지를 던지도록 변경했습니다!

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,72 @@
+    public void move(int x, int y, int destinationX, int destinationY) {
+        Position source = new Position(x, y);
+        Position destination = new Position(destinationX, destinationY);
+        List<Piece> existingPieces = new ArrayList<>(pieceMap.values().stream().toList());
```
`stream().toList()`처리했는데 다시 `new ArrayList`로 감싸는 이유가 궁금해요~

cookie-meringue replied:
지금은 없는 구조이지만, 당시 상황에서는 `existingPieces` 에 움직일 `Piece` 를 제외한 모든 `Piece` 들을 넣어야 했습니다.

이 과정에서, 모든 `Piece` 들에서 움직일 `Piece` 를 빼 주어야 했는데요

`remove()` 를  수행하기 위해 `new ArrayList` 로 감쌌습니다. `pieceMap.values().stream().toList()` 은 불변 리스트를 반환하기 때문입니다.

이 방식을 사용하면 시간복잡도가 O(n) + O(n) 이므로, 좋은 방식이 아니라는 것은 인지하고 있었습니다!

sinb57 commented on src/main/java/janggi/controller/Controller.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            Entry<Integer, Integer> source = inputView.inputHanMoveSource();
+            Entry<Integer, Integer> destination = inputView.inputHanMoveDestination();
+            janggiBoard.move(source.getKey(), source.getValue(), destination.getKey(), destination.getValue());
```
getter 사용은 지양해주세요~

cookie-meringue replied:
해당 getter 는 Entry 에 대한 getter 였습니다!
최대한 빠르게 구현하기 위해 따로 클래스를 만들지 않고 Entry 를 사용했는데요, 애초에 Entry 자체를 사용하지 않도록 변경하겠습니다.

sinb57 replied:
Entry 였군요! 코드를 잘못 봤네요,, 그럼에도 Entry 사용을 돌이켜보고 수정해주셔서 감사해요~

sinb57 replied:
아 코드를 잘못 봤네요!! Entry에 대한 getter군요😂

sinb57 replied:
아직 코드상에 Entry를 사용중인데, Entry 사용을 유지한 이유가 있나요? Position 객체를 생성해도 좋아보여요

sinb57 commented on src/main/java/janggi/domain/piece/Piece.java:
```diff
@@ -0,0 +1,84 @@
+
+public abstract class Piece {
+
+    private final Side side;
+    private Position position;
```
private으로 접근 제한할 필요가 있을까요?

cookie-meringue replied:
지금은 이 방식으로 구현하지 않아 리뷰 반영은 어렵지만, 이 방식에 대해 궁금한 것이 있어 PR 본문으로 남기겠습니다.
꼼꼼하게 봐 주셔서 감사합니다.

sinb57 commented on README.md:
```diff
@@ -1,3 +1,45 @@
+
+3. 장기판
+    - [x] 각 진영 별로 기물들의 위치를 초기화시킨다.
+    - 각 진영 별로 기물을 움직일 수 있다.
```
실행해보니 어떤 입력을 해야할지 막막했어요. 어디가 0,0 인지 기준 잡기 어렵더라고요. 장기 판의 x축과 y축에 좌표를 보여주면 어떨까요? 아래 사진 속 체스판 처럼요.

<img src="https://github.com/user-attachments/assets/8399ea1a-42ed-42fa-afb8-09f7498ee5ca" width=400>


그리고 README 가 사용 설명서의 역할을 수행할 수 있도록, 내용을 보충해주시면 좋겠어요~ 입력 예시를 추가해보면 어떨까요?

cookie-meringue replied:
사용 설명서 작성했습니다!

sinb57 reviewed (changes requested):
안녕하세요 머랭! 만나서 반가워요ㅎ Level1 마지막 미션의 리뷰어를 맡은 이프 예요!
장기 미션 하시느라 고생 많으셨어요. 첫번째 리뷰 요청 기한이 빠듯했죠?ㅠ
코드 살펴봤는데, 구현 및 설계가 완전치 않아서 리뷰 드릴 내용이 많지 않네요..
마저 완성하신 후 요청주시면 다시 봐드릴게요!

cookie-meringue committed:
refactor: Piece 맵 변수명 변경

cookie-meringue committed:
refactor: 나이트 이동 로직 최적화

cookie-meringue committed:
refactor: 폰 이동 로직 최적화

cookie-meringue committed:
refactor: 룩 이동 로직 최적화

cookie-meringue committed:
refactor: 이동 관련 함수명을 명확하게 변경

cookie-meringue committed:
feat: 방향 구현

cookie-meringue committed:
feat: 벡터 구현

cookie-meringue committed:
feat: Piece 일급 컬렉션 구현

cookie-meringue committed:
feat: 움직임 전략 인터페이스 추가

cookie-meringue committed:
feat: 고정된 움직임 전략 인터페이스 추가

cookie-meringue committed:
feat: 엘리펀트 움직임 전략 인터페이스 구현

cookie-meringue committed:
refactor: 엘리펀트 움직임 전략을 사용하도록 변경

cookie-meringue committed:
remove: 필요 없는 주석 제거

cookie-meringue committed:
refactor: 가드의 움직임을 인터페이스로 분리

cookie-meringue committed:
refactor: 킹의 움직임을 인터페이스로 분리

cookie-meringue committed:
refactor: 상수명을 Vector 로 변경

cookie-meringue committed:
refactor: 사용하지 않는 상수와 함수 제거

cookie-meringue committed:
refactor: 나이트의 움직임을 인터페이스로 분리

cookie-meringue committed:
refactor: 매개변수를 Collection 으로 추상화

cookie-meringue committed:
feat: 동적 움직임 전략 내부 구현

cookie-meringue committed:
refactor: 폰 이동 로직 전략 구현

cookie-meringue committed:
refactor: 룩 이동 로직 전략 구현

cookie-meringue committed:
refactor: 말이 움직임 전략을 주입받아 사용하도록 변경

cookie-meringue committed:
refactor: 피스의 움직임 책임을 보드로 전가

cookie-meringue committed:
refactor: 보드 내 로직 최적화

cookie-meringue committed:
refactor: Side 패키지 변경

cookie-meringue committed:
style: 코드 포맷 적용

cookie-meringue commented:
제 페어가 실수로 리뷰 반영을 제 레포지토리에 push 해서 덮어쓰기 위해 force push 를 하였습니다.

cookie-meringue committed:
docs: 사용 설명서 작성

cookie-meringue committed:
refactor: 의미 있는 변수를 사용하도록 변경

cookie-meringue committed:
refactor: 움직임 전략들을 추상 클래스 구현체로 변경

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,64 @@
+        placedPieces = Pieces.from(Stream.concat(hanPieces.stream(), choPieces.stream()).collect(Collectors.toList()));
+    }
+
+    public void move(int x, int y, int destinationX, int destinationY) {
```
position 타입으로 인자를 받으면 어떨까요?

cookie-meringue replied:
`Position` 은 `Piece` 와 관련된 도메인이라고 생각했습니다.
`Piece` 의 움직임에서 사용되기 때문입니다.

이 관점에서 생각해봤을 때, `JanggiBoard` 가 `Position` 에 의존하는 것이 옳지 못하다는 결론을 내렸습니다.
`JanggiBoard` 가 `Position` 을 의존하게 되면, `Position` 의 변경이 `JanggiBoard` 까지 전파되기 때문입니다.

이러한 제 설계를 반영하기 위해 `piece` 패키지 외부에서 `Position` 을 의존하지 않도록 설계했습니다.
(`OutputView` 에서의 의존은 `DTO` 를 만들지 않는 이상 어쩔 수 없다고 판단했습니다)

링크는 어떻게 생각하시나요?

sinb57 replied:
제 생각과는 달랐네요. `Position`을 어디에도 의존하지 않은 순수한 값객체로 보았어요. 때문에 `int x, int y`를 개별 인자로 계속 전달하지 않고 값객체의 의도대로 `x,y`를 `Position`으로 감싸길 바랬어요.

인자 관계를 생각하였을 때 Position으로 응집시키면 인자의 개수를 줄일 수 있다고 생각했고요.
메서드에 인자가 잘못 들어갈 문제도 막을 수 있다고 생각했어요.
```java
// x,y 인자 순서를 뒤섞은 실수
board.move(x, y, destinationY, destinationX)

// x 인자들의 순서를 뒤섞은 실수
board.move(dX, y, x, destinationX)
}
```

4개의 인자가 모두 int 형이기 때문에 위와 같은 문제가 충분히 이뤄날 수 있다고 생각했고, 값객체의 여러 장점(응집, 타입)을 살려서 개선하길 바랬어요.

```java
void Tobe() {
    var from = Position(x, y)
    var to = Position(destinationX, destinationY)
    board.move(from, to)
}
```

sinb57 replied:
제 생각과는 달랐네요. `Position`을 어디에도 의존하지 않은 순수한 값객체로 보았어요. 때문에 `int x, int y`를 개별 인자로 계속 전달하지 않고 값객체의 의도대로 `x,y`를 `Position`으로 감싸길 바랬어요.

인자 관계를 생각하였을 때 Position으로 응집시키면 인자의 개수를 줄일 수 있다고 생각했고요.
메서드에 인자가 잘못 들어갈 문제도 막을 수 있다고 생각했어요.
```java
// x,y 인자 순서를 뒤섞은 실수
board.move(x, y, destinationY, destinationX)

// x 인자들의 순서를 뒤섞은 실수
board.move(dX, y, x, destinationX)
}
```

4개의 인자가 모두 int 형이기 때문에 위와 같은 문제가 충분히 이뤄날 수 있다고 생각했고, 값객체의 여러 장점(응집, 타입)을 살려서 개선하길 바랬어요.

```java
void Tobe() {
    var from = Position(x, y)
    var to = Position(destinationX, destinationY)
    board.move(from, to)
}
```

cookie-meringue replied:
외부에서 사용하는 관점에서는 생각해보지 못했네요.
말씀을 들어 보니, 외부에서 Position 을 통해 소통하는 것도 좋은 것 같습니다.
단순한 value object 가 아니라 애플리케이션 전역에서 고루 사용되는 value object 이기 때문이라는 생각도 드네요
감사합니다!

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,64 @@
+        placedPieces.put(sourcePiece);
+    }
+
+    private void validateSamePosition(int x, int y, int destinationX, int destinationY) {
+        if (x == destinationX && y == destinationY) {
+            throw new IllegalArgumentException("현재 위치로 이동할 수 없습니다.");
+        }
+    }
```
Position 객체를 활용하면 좋겠어요.

cookie-meringue replied:
https://github.com/woowacourse/java-janggi/pull/69/files#r2012576996

위 리뷰와 이어지는 내용이라 생각합니다. 만약 JanggiBoard 에서 Position 을 의존하게 된다면 Position 객체를 활용하면 좋을 것 같습니다.

sinb57 commented on src/main/java/janggi/domain/board/JanggiBoard.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public Map<Position, Piece> getPlacedPieces() {
+        return placedPieces.getValues();
+    }
```
불변 처리하는게 좋아보여요

cookie-meringue replied:
인정합니다!
방어적 복사와 clone 을 사용했습니다.

sinb57 replied:
Cloneable 인터페이스를 구현토록 했군요! 음 깊은 복사와 얕은 복사에 대해 알아보시겠어요?

cookie-meringue replied:
`clone()` 은 얕은 복사를 수행합니다.
`placedPieces` 의 내부 필드를 얕은 복사한 새로운 `placedPieces` 가 복제본으로 탄생됩니다.

이 상황에서, 두 `Pieces` 는 다른 참조를 가지지만, 내부 `Map` 의 주소는 같습니다.

그렇다면 왜 `clone()` 을 사용하는가? 에 대한 의문이 들 수 있다고 생각합니다.
저는 `Pieces` 에 `Map` 만이 존재한다는 가정을 하지 않았습니다.
만약 `Pieces` 가 참조 타입이 아닌 원시 타입도 필드로 가지게 된다면 어떻게 될까? 하는 생각을 했습니다.

이 상황에서 만약 `clone()` 이나 기타 복사생성자 등을 구현하지 않았다면, 리팩터링해야 할 것들이 더 많을 것이라 생각했습니다.
현 시점에서 보면 오버엔지니어링이라고 볼 수 있을 것 같습니다. (학습의 목적도 있었습니다)

sinb57 commented on src/main/java/janggi/domain/piece/movement/fixed/ElephantMovementStrategy.java:
```diff
@@ -0,0 +1,58 @@
+        Position origin,
+        Position destination,
+        List<Vector> movePatterns
+    ) {
+
+        return movePatterns.stream()
```
이런 줄바꿈이 여러 곳에서 보이는데, 의도된 걸까요?

cookie-meringue replied:
매개변수가 많아서 텍스트가 길어지면 위와 같은 형식으로 정렬하는 편입니다!

이렇게 하는 경우는 거의 없을까요? 이 정렬 방식에 대해 궁금하긴 했습니다!

sinb57 replied:
개인의 취향이지만, 네.. 불필요한 줄바꿈으로 보여요

cookie-meringue replied:
다음 리뷰 요청에서는 지워놓겠습니다!
코드를 읽는 사람 입장도 생각해야 한다고 생각해서입니다!

sinb57 commented on README.md:
```diff
@@ -1,3 +1,45 @@
+2. 기물을 생성한다.
+    - 기물은 장기판 내에 위치할 수 있다.
+
+3. 기물은 움직일 수 있다.
+    - 기물은 움직일 수 있는 방향이 정해져 있다.
```
가로막히는 경우 등 이동이 불가능한 케이스 처리가 안되어 있네요
<img width="358" alt="image" src="https://github.com/user-attachments/assets/c28da6de-12f4-4282-a8a0-8eea17a8b7b1" />

cookie-meringue replied:
리팩터링 도중에 코드가 빠졌네요
수정했습니다!

sinb57 commented on src/main/java/janggi/domain/piece/gererator/ChoPieceGenerator.java:
```diff
@@ -0,0 +1,9 @@
+import janggi.domain.piece.Piece;
+import java.util.List;
+
+public interface ChoPieceGenerator {
```
인터페이스를 만든 의도와, 이를 통해 얻은 이점이 무엇인지 궁금해요~

cookie-meringue replied:
OCP 를 준수하기 위해 인터페이스로 설계했습니다.

만약 각 나라 별로 기물 배치 템플릿이 여러 개가 생긴다면, 기존 클래스를 수정하는 것이 아니라 새로운 Generator 를 만들면 되기 때문입니다.

이와 더불어서, 런타임에 동적으로 구현체를 정할 수 있다는 장점도 있습니다.

그렇지만, 지금 상황에서는 얻을 수 있는 이점이 없다고 생각합니다.
미래에 얻을 수 있는 이점을 한 가지 제시해보겠습니다.
### 만약 사용자의 선택에 따라 초나라와 한나라의 기물 위치가 바뀐다고 가정
- 초나라의 기물이 아래로 와야 한다면, 초나라 피스를 아래에 배치하는 ChoPieceGenerator 를 만들어주면 됩니다.

sinb57 replied:
변화에 대응하는 장점을 지녔지만 현재 상황에 맞지 않기 때문에, 오버 엔지니어링으로 보이네요. 선택은 머랭의 몫이지만, 우테코 당시 확장을 매우 고려하며 오버 엔지니어링을 많이 했었어요. 그때 겪은 바와 현업에서 겪은 바를 짧게나마 공유드려요

클래스가 많아지고 상속과 인터페이스가 많아지면, 코드를 파악하는 입장에서는 숨겨진 의도를 해석하기 위해 많은 시간과 노력을 할애하게 돼요. 그래서 변경에 대응하기 위한 코드도 중요하지만, 현재에 충실한 코드가 무엇보다 중요해요. 필드에 나오시면 겪겠지만, 변경에 닫힌 코드도 그렇고 미래를 내다본 코드도 모두 힘들어요.

만약 미래의 변경을 대비한 열린 설계를 하고자 한다면, 무조건 우선시 되어야 하는 것이 기획자와의 소통이에요. 기획의 방향성과 프로덕트의 다음 과제 등을 같이 의논하고 전해들으면서 지금의 코드에 녹이는 것이 좋고, 그렇지 않은 상태로 추측성 확장은 오히려 그 방향으로 나아가지 않을 가능성이 높아요. 실제 현업은 사용자를 위해 내놓은 기능이 오히려 반응이 좋지 않아 없애기도 하고, 실험적으로 내놓은 기능이 대박을 쳐서 끊임없이 확장되어 별도의 프로덕트로 분리되기도 해요. 따라서 미래를 내다본 설계는 어느때보다도 신중해야 해요.

예시로 들어주신 기물 위치 변경은, 현재 비즈니스 로직을 그대로 둔채 UI 로직을 수정해서 반영할 수 있겠네요. 비즈니스 로직을 건드리면 어떤 사이드 이펙트가 발생할 지 몰라요. 되도록 건들지 않고, 그 외 방법으로 해결할 수는 없을지 고민하는 것도 중요해요.

cookie-meringue replied:
감사합니다!
너무 앞을 내다본 코드가 읽기 힘든 코드가 될 수도 있겠네요.
다음 리뷰 요청 때 수정하겠습니다!

sinb57 commented on src/main/java/janggi/domain/piece/movement/fixed/Vector.java:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    boolean hasRelativeOffsetFrom(Position origin, Position destination) {
+        int xDistance = origin.getXDistance(destination);
+        int yDistance = origin.getYDistance(destination);
+        return this.x() == xDistance && this.y() == yDistance;
+    }
```
distance를 가져와서 비교하지 말고, distance를 계산해서 비교하라고 메시지를 던지면 어떨까요? origin에게 destination과 this.x()를 같이 전달하는 거죠

cookie-meringue replied:
좋은 방법이라고 생각합니다. 다만, 우려되는 점은 있습니다.

만약 이프가 말씀하신대로 Position에 hasRelativeOffsetFrom() 를 만드는 상황에서 장단점을 생각해 보았습니다.

장점1 : Vector 에서 Position 의 내부 구현을 잘 몰라도 됩니다.(추상화)
장점2 : 응집도가 높습니다. Position 이 알아서 자신의 역할을 수행하기 때문입니다.

단점: Position 에 Vector 만을 위한 메서드가 존재해야 합니다. Position 의 hasRelativeOffsetFrom() 은 Vector 내부에서만 사용되고, 함수 로직 상 Vector 만을 위한 메서드가 되리라 생각합니다.

저는 장점이 더 와닿았기 때문에, 이프의 피드백을 반영했습니다.

제 사고 흐름에서 잘못된 점이나 추가적인 코멘트가 있으신지 궁금합니다!

sinb57 replied:
장단점에 공감해요! 지금의 `hasRelativeOffsetFrom`는 Vector 만을 위한 메서드로 보이는데요. 원래 아래와 같이 생각하고서 제안드렸어요. 한번 장단점을 비교해보시겠어요?

```java
record Vector(int width, int height) {
    ...
    boolean hasRelativeOffsetFrom(Position origin, Position destination) {
        return origin.isWidthSameWith(destination, width) && origin.isHeightSameWith(destination, height);
    }
}

public record Position(int x, int y) {
    ...
    public boolean isWidthSameWith(Position other, int distance) {
        return this.widthFrom(other) == distance;
    }
}
```

cookie-meringue replied:
억지스러운 느낌이 아니고 자연스러운 느낌이 되었네요!

감사합니다! 다음 리뷰 요청에 반영하겠습니다!

sinb57 commented on src/main/java/janggi/domain/piece/Position.java:
```diff
@@ -0,0 +1,70 @@
+
+public final class Position {
+
+    private static final Predicate<Integer> X_MOVEABLE = x -> x >= 0 && x < 9;
```
private 메서드로 분리해도 괜찮을 텐데, Predicate 타입의 상수로 처리한 이유가 궁금해요

cookie-meringue replied:
저는 생각의 흐름을 작성할 때 `Predicate` 로 작성하면 생각이 더 잘 정리됩니다.

`Position` 을 작성할 때에 리팩터링을 해야겠다고 남겨두었는데 빼먹은 것 같습니다!

가독성은 메서드를 통한 검증이 더 좋다고 생각합니다.

sinb57 commented on src/main/java/janggi/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,98 @@
+
+    public boolean containsPieceType(PieceType pieceType) {
+        return values.values().stream()
+            .anyMatch(piece -> piece.getPieceType().equals(pieceType));
```
getter 사용은 지양해주세요

cookie-meringue replied:
Pieces 와 소통하도록 변경했습니다!

(리팩터링이 덜 된 코드가 많네요 ㅎㅎ..)

sinb57 commented on src/main/java/janggi/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    public int countByPieceType(PieceType pieceType) {
+        return Math.toIntExact(
```
`Math.toIntExact(` 처리한 이유가 궁금해요~

cookie-meringue replied:
long 을 int 로 변환하는 방법은 두 가지가 있는 걸로 알고 있습니다.

1. `(int)` 로 형변환
2. `Math.toIntExtract()`


### `(int)` 로 형변환
장점: 가장 간단하고 가독성이 좋습니다.
단점: int overflow 를 처리할 수 없습니다.

### `Math.toIntExtract()`
장점: int overflow 가 발생하면 `ArithmeticException()` 을 던지기 때문에 안전한 형변환을 할 수 있습니다.
단점 1: 메서드를 한번 거치는 것이므로, 아주 작은 오버헤드가 존재합니다.
단점 2: 사람에 따라 다르겠지만, `(int)` 에 비해 가독성이 다소 떨어집니다.


이 두 방식을 비교했을 때, 저는 `Math.toIntExtract()` 를 습관하하자는 결정을 내렸습니다.

sinb57 replied:
음 왜 long을 int로 형변환해야 하는지 이해하진 못했어요. long 타입을 그대로 사용하면 안되나요???

cookie-meringue replied:
아 타입에 대해 말씀하시는거였군요!

저는 정수 자료형을 int 로 표현하는 편입니다!
늘 하던 대로 했던 것 같습니다.

그러나, 깊게 생각해 보니 장단점이 있는 것 같습니다.

장점 1: 메모리
- int 는 long 보다 적은 32 비트를 사용합니다. 따라서, long 보다 메모리 효율적입니다.

단점 1: 데이터가 많아지면?
- 데이터가 많아지면 long 으로 변경해야 할 수 있습니다(지금 시점엔 없을 것이라 예상됩니다)
- 그 시점에 리팩터링하기에 힘들 수 있습니다.
단점 2(추측): count 의 시그니처인 long 을 굳이 int 로 형변환
- count 의 반환 타입이 long 인데는 이유가 있으리라 생각합니다.
- 기본 반환 타입을 int 로 형변환함에 따른 문제가 있을 것 같은데 이 부분은 잘 떠오르지 않습니다.


이프는 어떤 의견이신가요?

sinb57 commented on src/main/java/janggi/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,98 @@
+        values.put(piece.getPosition(), piece);
+    }
+
+    public void removeByPosition(int x, int y) {
```
Position 클래스를 만들어두었는데 활용하지 않는 모습이 많이 보이네요. Position 객체를 주고받는게 좋아보이는데, x와 y를 주고받는 이유가 궁금해요~

cookie-meringue replied:
https://github.com/woowacourse/java-janggi/pull/69/files#r2012576996

위 리뷰와 이어지는 내용이라 생각합니다. 만약 JanggiBoard 에서 Position 을 의존하게 된다면 Position 객체를 활용하면 좋을 것 같습니다.

sinb57 commented on src/main/java/janggi/view/InputView.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Map.Entry;
+import java.util.Scanner;
+
+public class InputView {
```
view 패키지 안에 두 클래스가 있길래, 입력은 Inputview에서 이뤄지고 출력은 OutputView에서 이뤄진다고 생각했어요. 그런데 InputView에서 입력을 위한 안내 문구 출력이 이뤄지네요. InputView와 OutputView의 차이는 무엇인가요? View를 `Input / Output`으로 구분지을 필요가 있나요?

cookie-meringue replied:
> InputView와 OutputView의 차이는 무엇인가요? View를 Input / Output으로 구분지을 필요가 있나요?

`View` 를 구분지은 이유는 다음과 같습니다.
지금은 입력, 출력이 전부 `Console` 인 콘솔 애플리케이션입니다.
그러나, 확장성을 고려했을 때 입력, 출력 수단은 충분히 달라질 수 있다고 생각했습니다.
특히 장기 게임이기 때문에 UI 가 중요하기 때문입니다.

이 생각을 바탕으로 `View` 를 `InputView` 와 `OutputView` 로 분리하기로 결정했습니다.
제 설계를 반영하기 위해서는 `InputView`, `OutputView` 를 인터페이스로 만들어야 했습니다. 빠르게 구현하는 과정에서 인터페이스로 만드는 것을 후순위로 미루게 되었고, 곧 인터페이스로 리팩터링할 예정이었습니다.

> view 패키지 안에 두 클래스가 있길래, 입력은 Inputview에서 이뤄지고 출력은 OutputView에서 이뤄진다고 생각했어요. 그런데 InputView에서 입력을 위한 안내 문구 출력이 이뤄지네요.

엄밀히 따지자면, `InputView` 에서는 출력을 하면 안 됩니다.
`InputView` 는 입력만 담당하는 `View` 이고, `OutputView` 는 출력만을 담당하는 `View` 이기 때문입니다.
그렇지만, 입력을 받기 위해서는 무엇을 입력해야 하는지에 대한 출력도 필요하다는 생각을 했습니다.
만약 `입력 View` 에서 `Java Swing` 을 사용하고, `출력 View` 에서는 `Console` 을 사용하게 되었다고 가정하겠습니다.

이런 상황이 온다면 각 `Java Swing` 의 `Component`를 통해 입력을 하고 `Console` 에서 출력을 확인해야 합니다.
그렇게 되면, 사용자는 `Console` 에서 출력된 `"이동 위치를 입력해주세요"` 문구를 보고 다시 `Java Swing`의 `Component` 에 입력을 해야 합니다.

이는 사용자 경험을 나쁘게 만드는 설계라고 생각했습니다.

그에 따라, `InputView` 가 입력을 받기 위한 최소한의 출력도 수행하도록 설계했습니다.

sinb57 replied:
프로젝트가 커지고 입출력이 많아지면, InputView와 OutputView가 비대해지고 관리가 어려워질텐데요. 입력과 출력이 아닌, 화면을 기준으로 View를 분리하는 것은 어떻게 생각하시나요? 여러 화면을 가진 키오스크나, 여러 페이지를 지닌 웹 사이트 처럼요.

cookie-meringue replied:
좋은 것 같습니다! 화면을 기준으로 분리할 생각을 못했네요!

다음 리뷰 요청에 반영하겠습니다!

sinb57 commented on src/main/java/janggi/view/InputView.java:
```diff
@@ -0,0 +1,55 @@
+        return KnightElephantSettingCommand.fromCommand(SCANNER.nextLine());
+    }
+
+    public Entry<Integer, Integer> inputHanMoveSource() {
+        System.out.println("한나라의 움직일 말의 위치를 입력하세요. (예: 1,2) : ");
+        return inputPosition();
+    }
```
Position 객체를 반환하면 안되나요?

cookie-meringue replied:
저는 `View` 에서 `model` 에 의존하는 것을 선호하지 않았습니다!
`View` 와 `Model` 을 연결하는 역할을 `Controller` 에게 위임하고 `DTO` 를 사용하는 것을 선호합니다.
물론, 이 방식도 결국 `View` 가 간접적으로 `Model` 에 의존한다는 문제는 있다고 생각합니다.

이러한 제 생각을 반영하기 위해 당시엔 `Entry<Integer, Integer>` 를 사용했지만, 좋은 방식이 아니라고 생각합니다.
**좋은 방식이 아닌 이유1**: 특정 자료구조에 종속됩니다. 만약 `x, y, z` 를 입력받게 된다면 리팩터링에 리소스가 많이 듭니다.
**좋음 방식이 아닌 이유2**: `Entry`는 이런 방식으로 사용되기 위해 존재하는 인터페이스가 아닙니다. `key - value` 한 쌍을 저장하기 위한 인터페이스로, 의도와 다르게 사용했다고 생각합니다.

그럼에도 제가 지금까지 `Entry` 를 사용하도록 유지한 이유는 `model` 패키지 내부에 초점을 맞추고 싶었기 때문입니다.
- `view`, `controller` 는 `model`과 비교해 후순위라는 생각을 했습니다. 객체지향 설계를 학습한다는 관점에서는 `model` 내부 도메인들간의 관계, 역할을 잘 설계하는 것이 가장 중요하다고 생각했기 때문입니다.

그러나, 이미 `OutputView` 에서 `Model` 을 직접적으로 강하게 의존하고 있기 때문에, `Entry` 보다는 `Position` 을 반환하는 것이 더 좋을 것이라는 생각이 듭니다.

반영하겠습니다!

sinb57 replied:
Controller와 DTO 얘기가 나와서 질문드려보아요~ 비즈니스 로직과 UI 로직을 구분지어야 하는 이유는 무엇일까요? 비즈니스 로직과 UI 로직을 구분 짓는 이유는 단순히 서로 몰라야 한다가 아니에요. 정말 서로가 서로를 모를 수 있을까요? 이런 흐름이 결국 Controller를 만들어내곤 하는데요. 비즈니스 로직과 UI 로직을 구분 짓고자 하는 이유는 무엇일까요? 정말 비즈니스 로직이 도메인을 몰라야 하나요?

그리고 DTO는 무엇일까요? View와 Model의 계층 간 전달을 위한 객체인가요??

cookie-meringue replied:
> 비즈니스 로직과 UI 로직을 구분지어야 하는 이유는 무엇일까요?

유지보수성 때문이라고 생각합니다. 저는 지금까지 View 와 Model 을 분리하고 의존성을 분리하는 과정에서 이 점을 많이 느꼈습니다. 출력 로직이 변경되었을 때 View 만 고치는 상황이 그 예시입니다!
"몰라야 된다" 라는 표현은 이를 구현하기 위해 사용되는 표현이라고 생각합니다. 이 말의 저의는 **View 와 Model 의 의존성을 최소화하기 위해 노력해야 한다** 로 생각하고 있습니다.

굳이 Model, View, Controller 라는 네이밍으로 분리하지는 않아도 된다고 생각합니다.
역할 분리만 잘 되어 있으면 그만이라고 생각합니다.


> 정말 비즈니스 로직이 도메인을 몰라야 하나요?

비즈니스 로직이 담겨 있는 부분이 도메인, Model 이라고 생각합니다.

sinb57 commented on src/main/java/janggi/domain/piece/movement/dynamic/PawnMovementStrategy.java:
```diff
@@ -0,0 +1,52 @@
+        int x = origin.getX();
+        int y = origin.getY();
+        if (side == Side.HAN) {
+            return Set.of(new Position(x, y + MOVABLE_DISTANCE));
```
getter로 꺼내는 것이 아니라, Position에 MOVABLE_DISTANCE를 전달해서 스스로 객체를 생성토록 하면 좋겠어요~ Position의 값객체 성질을 잘 활용해보세요

cookie-meringue replied:
값 객체에 대해서 공부했었는데.. 막상 사용하지 못했네요

좋은 피드백 감사합니다!

sinb57 reviewed (changes requested):
안녕하세요 머랭~ 리팩터링하느라 고생 많으셨어요! 덕분에 코드 잘 읽었답니다ㅎ
추가 코멘트 남겼으니 확인 해주세요~

---

> 1. 상속 시 부모 필드에 접근하는 방법

getter를 통해 접근하면 되기 때문에 프로퍼티에 직접 접근하는 것을 막고자 했군요! 상황에 따라 다르고 정답은 없지만, 프로퍼티에 접근하는 것이 문제된다고 생각하지 않아요. `자식 클래스에 부모의 Position 을 null 로 만드는 등의 잘못된 접근을 하면 위험하지 않을까?`라는 걱정에 대해서는, `final` 선언으로 변경을 막으면 되니까요. 오히려 getter가 필요없는 순간임에도 프로퍼티에 접근하기 위해 하위 클래스들에게 getter를 열어주는 것은, 직진하면 되는데 굳이 길을 돌아가는 느낌이네요.

> 2. 기물의 움직임을 전략 인터페이스로 추상화

잘 해주셨어요! 이동 패턴에 기반해서 계층을 구분지으셨네요. 변화에 유연한 대처가 가능해졌기에 좋은 설계라고 봐요~

cookie-meringue committed:
fix: 룩 이동 시 목적지 사이에 기물이 존재하면 이동할 수 없도록 수정

cookie-meringue committed:
refactor: 장기 보드가 Position 에 의존하지 않도록 변경

cookie-meringue committed:
feat: 피스에 접근만 가능한 View 인터페이스 추가

cookie-meringue committed:
feat: Pieces 에 접근만 가능한 View 인터페이스 추가

cookie-meringue committed:
refactor: PieceView 와 PiecesView 를 사용하도록 변경

- 의도하지 않은 상태 변경을 방지하기 위함.

cookie-meringue committed:
refactor: 좌표 검증 시 Predicate 를 사용하지 않도록 변경

cookie-meringue committed:
refactor: 불변 클래스인 Position 을 Record 로 변경

cookie-meringue committed:
fix: 포지션 생성 시 x, y 값 검증 로직 수정

cookie-meringue committed:
refactor: 상, 마 생성 로직을 Enum 내부로 이동

cookie-meringue committed:
refactor: Position 의 getter 를 사용하지 않도록 변경

cookie-meringue committed:
refactor: 벡터 계산 시 Position 에게 메시지를 던지도록 변경

cookie-meringue committed:
refactor: 좌표 입력 시 InputView 에서 Position 을 반환하도록 변경

cookie-meringue committed:
refactor: getter 에서 방어적 복사를 수행하도록 변경

- 수정 불가 Map 으로 반환하도록 변경.
- 내부의 Entry 수정이 외부에 반영되지만, View 에서만 사용되기 때문에, 복사해서 반환하지는 않도록.

cookie-meringue committed:
rename: 잘못된 패키지명 수정

- gererator -> generator.

cookie-meringue committed:
fix: King 움직임 테스트에서 King 움직임 전략을 사용하도록 수정

cookie-meringue committed:
rename: 함수명 오타 수정

cookie-meringue committed:
refactor: 고정 움직임 전략 로직 수정

- 중간 기물들도 확인하도록 변경.

cookie-meringue committed:
refactor: 캐논 테스트 케이스 수정

sinb57 commented on src/main/java/janggi/domain/piece/Position.java:
```diff
@@ -0,0 +1,70 @@
+        validate(x, y);
+    }
+
+    private void validate(int x, int y) {
+        if (x < X_MIN_VALUE || x > X_MAX_VALUE) {
+            throw new IllegalArgumentException("x의 범위가 잘못되었습니다.");
+        }
+        if (y < Y_MIN_VALUE || y > Y_MAX_VALUE) {
+            throw new IllegalArgumentException("y의 범위가 잘못되었습니다.");
+        }
+    }
```
`int x, int y` 인자타입이 서로 같기 때문에, 자칫 인자 순서를 잘못 전달하는 실수가 발생할 수 있어보여요. x와 y 원시값을 포장해보면 어떨까요?

더불어 검증에 대한 책임도 각 객체에 위임할 수 있겠네요

sinb57 reviewed (approved):
안녕하세요 머랭~ 2단계 진행을 위해 머지할게요! 잘 해주셨어요ㅎ
추가 코멘트 확인해주시고 2단계 PR에 답변 부탁드려요
